job-runner/chain/traverser.go

```diff
@@ -377,3 +395,47 @@ func (t *traverser) sendJL(pJob proto.Job, err error) {
 		jLogger.Errorf("problem sending job log (%#v) to the RM: %s", jl, err)
 	}
 }
+
+// retrySequence retries a sequence by doing the following
+//   1. Mark the failed job and all jobs before it in the sequence to PENDING
+//   state, so that they are ran again on sequence retry. This is OK to do
+//   since all jobs are idempotent.
```

Comment:
Fwiw, there is currently no mention of jobs being idempotent in the design doc. I know we generally try to make our jobs that way, but I'm not sure we've officially agreed that all jobs created for spincycle should be idempotent

Reply:
Hm, ok. Thinking about this a bit and there are a few options:

1) when we configure a sequence retry, ensure that all jobs it may affect are idempotent. However, someone may forget to check or we may later add a job to the retryable sequence that's not idempotent. Given that nothing can be ensured, I don't really like this best effort / check as we go approach. We must be doing this approach for job retries, though.
2) Make it a requirement that all jobs are idempotent. This would require doing an audit before merging in this change and updating the design docs. I'm more of a fan of this one.

WDYT?

Reply:
I think this is a documentation issue for elsewhere, isn't a matter for this private function. For user-facing docs, we should say that jobs _should_ support being retried (being truly idempotent is a strong requirement; is possible to support retry with slightly different results on retry, which means it's not technical idempotent). But really it's up to user because it's not something Spin Cycle itself can ever check or enforce. We just run the jobs... up to user to write code to handle retries if they use retries.